src/sentry/releasehealth/base.py

```diff
@@ -100,3 +108,16 @@ def get_release_adoption(
         """
 
         raise NotImplementedError()
+
+    def check_has_health_data(
+        self, projects_list: Sequence[Union[ProjectId, ProjectRelease]]
```

Comment:
There are two problems with this type:

* There's no relation between input and output.. type signature implies that one can pass in releases, and get out just projects. Or that one passes in projects and gets out releases (sometimes?)
* Union at this position implies that you can freely mix `ProjectId` and `ProjectRelease` within the same sequence but the implementation doesn't handle that

You can make this generic over a `T` that is bounded by `Union[ProjectId, ProjectRelease]`. Then you have the guarantee that the same union variant is returned as is put into it.

```
T = TypeVar("ProjectOrRelease", bounds=Union[...])

def check_has_health_data(self, projects_list: Sequence[T]) -> Set[T]: ...
```

This will make the function generic over `T: Union[...]`. You should probably pick a better name for `T` though.

Reply:
Thanks for the input, will do.

From what I understand making `T=TypeVar("T", bounds=Union[ProjectId,ProjectRelease])` would not improve the semantics in any way, but making the definition `T=TypeVar("T", bounds= ProjectId, ProjectRelease)` would do exactly what you suggest.

Reply:
Right, that may be better. However in any case it will be better, because when you call `check_has_health_data` you can declare `T` to be specifically `ProjectId` either way